third/nmh/man/rcvtty.man
.\"
.\" %nmhwarning%
.\" $Id: rcvtty.man,v 1.1.1.1 1999-02-07 18:14:22 danw Exp $
.\"
.\" include the -mh macro file
.so %etcdir%/tmac.h
.\"
.TH RCVTTY %manext1% MH.6.8 [%nmhversion%]
.SH NAME
rcvtty  \- report new mail
.SH SYNOPSIS
.in +.5i
.ti -.5i
%libdir%/rcvtty
\%[command]
.br
\%[\-form\ formatfile]
\%[\-format\ string]
\%[\-width\ columns]
.br
\%[\-bell] \%[\-nobell]
\%[\-newline]
\%[\-nonewline]
\%[\-biff]
.br
\%[\-version]
\%[\-help]
.in -.5i
.SH DESCRIPTION
The \fIrcvtty\fR can be used to report new mail.  It is used primarily
in conjunction with mail filtering agents such as \fIslocal\fP or
\fIprocmail\fP.

The \fIrcvtty\fR program executes the named command with the message as
its standard input, and writes the resulting output on your terminal.

Alternately, if no command is specified (or is bogus), then \fIrcvtty\fR
will instead write a one\-line scan listing.  The default output format
of this scan listing may be overridden by using either the
`\-form\ formatfile' or `\-format\ string' option, similar to the
equivalent options for \fIscan\fP and \fIinc\fP.
See \fImh\-format\fP(5) for details.

A newline is output before the message output, and the terminal bell is
rung after the output.  The `\-nonewline' and `\-nobell' options will
inhibit these functions.

The switch `\-width\ columns' may be given to specify the width of
the scan line.  The default is to use the width of the terminal.

In addition to the standard format escapes described in
\fImh\-format\fR(5), \fIrcvtty\fR also recognizes the following additional
\fIcomponent\fR escapes:
.sp 1
.ne 5
.nf
.ta \w'Dtimenow  'u +\w'Returns  'u
\fIEscape\fR	\fIReturns\fR	\fIDescription\fR
body	string	the (compressed) first part of the body
dtimenow	date	the current date
folder	string	the name of the current folder
.re
.fi

By default, \fIrcvtty\fP will send its output to every terminal on the
local machine that is owned by current user, and that has given write
permission as granted by the command \fImesg\fP\0(1).  If the option
`\-biff' is given, then \fIrcvtty\fP will obey the notification status
set by the command \fIbiff\fP\0(1) instead.
.Fi
^%etcdir%/mts.conf~^nmh mts configuration file
^$HOME/\&.maildelivery~^The file controlling local delivery
^%etcdir%/maildelivery~^Rather than the standard file
.De
`\-width' defaults to the width of the terminal
.Ds
`\-newline'
.Ds
`\-bell'
.Sa
rcvdist(1), rcvpack(1), rcvstore(1), mh\-format(5), slocal(1)
.Bu
Only two return codes are meaningful, others should be.
.En


doc/news
This file is intended to keep track of user-visible changes to the
release, for the construction of the Athena release notes.

Changes from release 8.0 to release 8.1:

	* Athena version numbers are now dotted triplets, e.g. 8.1.0
	  or 8.1.12.  Paths which include all or part of the Athena
	  version number now take the form "pathname-version",
	  e.g. "srvd-8.1" or "dash.menus-8.1"

	* In release 8.0, running "xmkmf" with the default path would
	  generate a Makefile which looked for its include files and
	  libraries in /mit/x11, which could be one of several
	  lockers.  Now it is preferrable to use the vendor X
	  libraries, so this bit of glue is going away.  There will be
	  no /usr/athena/bin/xmkmf any more, nor any symlinks in
	  /usr/athena pointing to the x11 locker.

	  Users can still do something like "add -f x11r5" and then
	  "xmkmf", and things should work, if the x11r5 locker is
	  built for their platform.

	* Athena bitmaps have moved to /usr/athena/include/X11/bitmaps.
	  Previously, they were in /usr/athena/lib/X11/bitmaps before
	  because of the /mit/x11 glue.

	* The symlink /bin/csh -> /bin/athena/tcsh has gone away.  The
	  update to 8.0 will fix /bin/csh references in /etc/passwd
	  and /etc/passwd.local to refer to /bin/athena/tcsh instead.

	* emacs18 has gone away.

	* AFS includes and libraries are installed in
	  /usr/afsws/include and /usr/afsws/lib.

	* wcl includes and libraries are installed in /usr/athena.

	* The basic krb5 headers and libraries are installed in
	  /usr/athena.

	* emacs has been upgraded from 19.30 to 19.34.  There should
	  be no significant user-visible changes.

	* RCS has been upgraded from version 4 to version 5.7.

	* x3270 has been upgraded from version 1.2 to version 3.1.1.3.

	* sysinfo is now available under Irix.

	* /bin/athena/tcsh now has Hesiod support enabled, which means
	  that "~lockername" will work in a pathname in most cases.
	  (It won't work for users with local passwd information who
	  don't have their lockers attached.)

	* For improved ease of debugging, many binaries are installed
	  with symbol tables, and compiled without optimization.
	  Therefore, the contents of /bin/athena and /usr/athena/bin
	  may be somewhat larger, and the binaries may consume more
	  CPU cycles, although it's unlikely that anyone will notice
	  the speed difference.


third/nmh/uip/rcvtty.c

/*
 * rcvtty.c -- a rcvmail program (a lot like rcvalert) handling IPC ttys
 *
 * $Id: rcvtty.c,v 1.1.1.1 1999-02-07 18:14:16 danw Exp $
 */

#include <h/mh.h>
#include <h/signals.h>
#include <h/rcvmail.h>
#include <h/scansbr.h>
#include <zotnet/tws/tws.h>
#include <signal.h>
#include <fcntl.h>

#include <utmp.h>
#ifndef UTMP_FILE
# ifdef _PATH_UTMP
#  define UTMP_FILE _PATH_UTMP
# else
#  define UTMP_FILE "/etc/utmp"
# endif
#endif

#define	SCANFMT	\
"%2(hour{dtimenow}):%02(min{dtimenow}): %<(size)%5(size) %>%<{encrypted}E%>\
%<(mymbox{from})%<{to}To:%14(friendly{to})%>%>%<(zero)%17(friendly{from})%>  \
%{subject}%<{body}<<%{body}>>%>"

static struct swit switches[] = {
#define	BIFFSW	0
    { "biff", 0 },
#define	FORMSW	1
    { "form formatfile", 0 },
#define	FMTSW	2
    { "format string", 5 },
#define WIDTHSW 3
    { "width columns", 0 },
#define NLSW    4
    { "newline", 0 },
#define NNLSW   5
    { "nonewline", 0 },
#define BELSW	6
    { "bell", 0 },
#define	NBELSW	7
    { "nobell", 0 },
#define VERSIONSW 8
    { "version", 0 },
#define	HELPSW	9
    { "help", 4 },
    { NULL, 0 }
};

static jmp_buf myctx;
static int bell = 1;
static int newline = 1;
static int biff = 0;
static int width = 0;
static char *form = NULL;
static char *format = NULL;

/*
 * external prototypes
 */
char *getusername(void);

/*
 * static prototypes
 */
static RETSIGTYPE alrmser (int);
static int message_fd (char **);
static int header_fd (void);
static void alert (char *, int);


int
main (int argc, char **argv)
{
    int md, vecp = 0;
    char *cp, *user, buf[BUFSIZ], tty[BUFSIZ];
    char **argp, **arguments, *vec[MAXARGS];
    struct utmp ut;
    register FILE *uf;

#ifdef LOCALE
    setlocale(LC_ALL, "");
#endif
    invo_name = r1bindex (argv[0], '/');

    /* read user profile/context */
    context_read();

    mts_init (invo_name);
    arguments = getarguments (invo_name, argc, argv, 1);
    argp = arguments;

    while ((cp = *argp++)) {
	if (*cp == '-') {
	    switch (smatch (++cp, switches)) {
		case AMBIGSW: 
		    ambigsw (cp, switches);
		    done (1);
		case UNKWNSW: 
		    vec[vecp++] = --cp;
		    continue;

		case HELPSW: 
		    snprintf (buf, sizeof(buf), "%s [command ...]", invo_name);
		    print_help (buf, switches, 1);
		    done (1);
		case VERSIONSW:
		    print_version(invo_name);
		    done (1);

		case BIFFSW:
		    biff = 1;
		    continue;

		case FORMSW: 
		    if (!(form = *argp++) || *form == '-')
			adios (NULL, "missing argument to %s", argp[-2]);
		    format = NULL;
		    continue;
		case FMTSW: 
		    if (!(format = *argp++) || *format == '-')
			adios (NULL, "missing argument to %s", argp[-2]);
		    form = NULL;
		    continue;

		case WIDTHSW:
		    if (!(cp = *argp++) || *cp == '-')
			adios(NULL, "missing argument to %s", argp[-2]);
		    width = atoi(cp);
		    continue;
                case NLSW:
                    newline = 1;
                    continue;
                case NNLSW:
                    newline = 0;
                    continue;
                case BELSW:
                    bell = 1;
                    continue;
                case NBELSW:
                    bell = 0;
                    continue;

	    }
	}
	vec[vecp++] = cp;
    }
    vec[vecp] = 0;

    if ((md = vecp ? message_fd (vec) : header_fd ()) == NOTOK)
	exit (RCV_MBX);

    user = getusername();
    if ((uf = fopen (UTMP_FILE, "r")) == NULL)
	exit (RCV_MBX);

    while (fread ((char *) &ut, sizeof(ut), 1, uf) == 1)
	if (ut.ut_name[0] != 0
		&& strncmp (user, ut.ut_name, sizeof(ut.ut_name)) == 0) {
	    strncpy (tty, ut.ut_line, sizeof(ut.ut_line));
	    alert (tty, md);
	}

    fclose (uf);
    exit (RCV_MOK);
}


static RETSIGTYPE
alrmser (int i)
{
#ifndef RELIABLE_SIGNALS
    SIGNAL (SIGALRM, alrmser);
#endif

    longjmp (myctx, 1);
}


static int
message_fd (char **vec)
{
    pid_t child_id;
    int bytes, fd, seconds;
    char tmpfil[BUFSIZ];
    struct stat st;

    unlink (mktemp (strncpy (tmpfil, "/tmp/rcvttyXXXXX", sizeof(tmpfil))));
    if ((fd = open (tmpfil, O_RDWR | O_CREAT | O_TRUNC, 0600)) == NOTOK)
	return header_fd ();
    unlink (tmpfil);

    if ((child_id = vfork()) == NOTOK) {
	/* fork error */
	close (fd);
	return header_fd ();
    } else if (child_id) {
	/* parent process */
	if (!setjmp (myctx)) {
	    SIGNAL (SIGALRM, alrmser);
	    bytes = fstat(fileno (stdin), &st) != NOTOK ? (int) st.st_size : 100;

	    /* amount of time to wait depends on message size */
	    if (bytes <= 100) {
		/* give at least 5 minutes */
		seconds = 300;
	    } else if (bytes >= 90000) {
		/* but 30 minutes should be long enough */
		seconds = 1800;
	    } else {
		seconds = (bytes / 60) + 300;
	    }
	    alarm ((unsigned int) seconds);
	    pidwait(child_id, OK);
	    alarm (0);

	    if (fstat (fd, &st) != NOTOK && st.st_size > (off_t) 0)
		return fd;
	} else {
	    /*
	     * Ruthlessly kill the child and anything
	     * else in its process group.
	     */
	    KILLPG(child_id, SIGKILL);
	}
	close (fd);
	return header_fd ();
    }

    /* child process */
    rewind (stdin);
    if (dup2 (fd, 1) == NOTOK || dup2 (fd, 2) == NOTOK)
	_exit (-1);
    closefds (3);
    setpgid ((pid_t) 0, getpid ());	/* put in own process group */
    execvp (vec[0], vec);
    _exit (-1);
}


static int
header_fd (void)
{
    int fd;
    char *nfs, tmpfil[BUFSIZ];

    strncpy (tmpfil, m_tmpfil (invo_name), sizeof(tmpfil));
    if ((fd = open (tmpfil, O_RDWR | O_CREAT | O_TRUNC, 0600)) == NOTOK)
	return NOTOK;
    unlink (tmpfil);

    rewind (stdin);

    /* get new format string */
    nfs = new_fs (form, format, SCANFMT);
    scan (stdin, 0, 0, nfs, width, 0, 0, NULL, 0L, 0);
    if (newline)
        write (fd, "\n\r", 2);
    write (fd, scanl, strlen (scanl));
    if (bell)
        write (fd, "\007", 1);

    return fd;
}


static void
alert (char *tty, int md)
{
    int i, td, mask;
    char buffer[BUFSIZ], ttyspec[BUFSIZ];
    struct stat st;

    snprintf (ttyspec, sizeof(ttyspec), "/dev/%s", tty);

    /*
     * The mask depends on whether we are checking for
     * write permission based on `biff' or `mesg'.
     */
    mask = biff ? S_IEXEC : (S_IWRITE >> 3);
    if (stat (ttyspec, &st) == NOTOK || (st.st_mode & mask) == 0)
	return;

    if (!setjmp (myctx)) {
	SIGNAL (SIGALRM, alrmser);
	alarm (2);
	td = open (ttyspec, O_WRONLY);
	alarm (0);
	if (td == NOTOK)
	    return;
    } else {
	alarm (0);
	return;
    }

    lseek (md, (off_t) 0, SEEK_SET);

    while ((i = read (md, buffer, sizeof(buffer))) > 0)
	if (write (td, buffer, i) != i)
	    break;

    close (td);
}

